File: backend/app/schemas/architecture.py
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


class GroupKind(str, Enum):
    TIRADA = "tirada"
    PLANO = "plano"
    FASE = "fase"


class ArchitectureItem(BaseModel):
    id: UUID
    descripcion: str = Field(..., min_length=1)
    capitulo: Optional[str] = None
    partida: Optional[str] = None
    unidad: Optional[str] = None
    cantidad: Optional[float] = Field(default=None, ge=0)
    precio_unitario: Optional[float] = Field(default=None, ge=0)
    subtotal: Optional[float] = Field(default=None, ge=0)
    notas: Optional[str] = None


class ArchitectureGroup(BaseModel):
    id: UUID
    kind: GroupKind
    title: str = Field(..., min_length=1)
    order: int = Field(..., ge=0)
    items: list[ArchitectureItem] = Field(default_factory=list)


class MaterialRow(BaseModel):
    id: UUID
    categoria: Optional[str] = None
    descripcion: str = Field(..., min_length=1)
    unidad: Optional[str] = None
    cantidad_estimada: Optional[float] = Field(default=None, ge=0)
    desperdicio_porcentaje: Optional[float] = Field(default=None, ge=0, le=100)
    cantidad_total: Optional[float] = Field(default=None, ge=0)
    costo_estimado: Optional[float] = Field(default=None, ge=0)
    proveedor_sugerido: Optional[str] = None


class ArchitectureDocumentPayload(BaseModel):
    groups: list[ArchitectureGroup] = Field(default_factory=list)
    materiales: list[MaterialRow] = Field(default_factory=list)

    @field_validator("groups")
    @classmethod
    def sort_groups(cls, v: list[ArchitectureGroup]) -> list[ArchitectureGroup]:
        return sorted(v, key=lambda g: g.order)


class ArchitectureDataResponse(BaseModel):
    project_uuid: UUID
    document: ArchitectureDocumentPayload
    updated_at: Optional[str] = None

File: frontend/src/schemas/architecture.ts
import { z } from 'zod'

export const groupKindSchema = z.enum(['tirada', 'plano', 'fase'])

export const architectureItemSchema = z.object({
  id: z.string().uuid(),
  descripcion: z.string().min(1),
  capitulo: z.string().optional().nullable(),
  partida: z.string().optional().nullable(),
  unidad: z.string().optional().nullable(),
  cantidad: z.number().nonnegative().optional().nullable(),
  precio_unitario: z.number().nonnegative().optional().nullable(),
  subtotal: z.number().nonnegative().optional().nullable(),
  notas: z.string().optional().nullable(),
})

export const architectureGroupSchema = z.object({
  id: z.string().uuid(),
  kind: groupKindSchema,
  title: z.string().min(1),
  order: z.number().int().nonnegative(),
  items: z.array(architectureItemSchema),
})

export const materialRowSchema = z.object({
  id: z.string().uuid(),
  categoria: z.string().optional().nullable(),
  descripcion: z.string().min(1),
  unidad: z.string().optional().nullable(),
  cantidad_estimada: z.number().nonnegative().optional().nullable(),
  desperdicio_porcentaje: z.number().min(0).max(100).optional().nullable(),
  cantidad_total: z.number().nonnegative().optional().nullable(),
  costo_estimado: z.number().nonnegative().optional().nullable(),
  proveedor_sugerido: z.string().optional().nullable(),
})

export const architecturePayloadSchema = z.object({
  groups: z.array(architectureGroupSchema),
  materiales: z.array(materialRowSchema),
})

export type ArchitecturePayload = z.infer<typeof architecturePayloadSchema>
